File: src/object_scout/src/object_scout/object_scanner.py
#!/usr/bin/env python3
import rospy
import math
from enum import Enum
from tf.transformations import quaternion_from_euler
from visualization_msgs.msg import Marker
from geometry_msgs.msg import Quaternion
from std_msgs.msg import Float32MultiArray

# Import actionlib for navigation
import actionlib


# ---------- ENUMERATIONS ----------

class ScanResult(Enum):
    """Enumeration for scan operation results"""
    NO_DETECTION = 0
    OBJECT_DETECTED = 1
    ERROR = 2

# ---------- CLASS DEFINITION ----------

class ObjectScanner:
    """
    Handles object detection through scanning rotations
    
    This class provides functionality to:
    1. Rotate the robot through a sequence of angles
    2. Detect objects at each rotation angle
    3. Verify detections with sustained observation
    4. Track scanning progress and remaining angles
    """
    def __init__(self, robot_name, nav_controller, object_mapper_topic,
                 bbox_depth_topic, init_node=False):
        """
        Initialize the scanner with robot name and navigation controller
        
        Args:
            robot_name (str): Name of the robot for topic namespacing
            nav_controller: Instance of NavigationController for rotation commands
            init_node (bool): Whether to initialize a ROS node (standalone mode)
        """
        # Initialize node if requested
        if init_node:
            rospy.init_node('object_scanner', anonymous=False)
        
        # ---------- CONFIGURATION ----------
            
        # Core components
        self.robot_name = robot_name
        self.nav_controller = nav_controller
        self.object_mapper_topic = object_mapper_topic
        self.bbox_depth_topic = bbox_depth_topic
        
        # Scanner parameters
        self.scan_stabilization_time = 1.0  # seconds
        self.required_detection_duration = rospy.Duration(1.0)  # seconds
        self.scan_timeout = rospy.Duration(1.5)  # seconds
        self.rotation_timeout = rospy.Duration(45.0)  # seconds
        self.max_detection_depth = 10.0  # meters
        
        # Default rotation angles in degrees
        self.rotation_angles = [0, 45, 90, 135, 180, 225, 270, 315]
        
        # ---------- STATE VARIABLES ----------
        
        # Scanner state
        self.scanning_in_progress = False
        self.object_marker = None
        self.object_detected = False
        self.current_depth = None
        self.current_class_id = None
        self.remaining_angles = []
        
        # ---------- ROS INTERFACE SETUP ----------
        
        # Set up marker subscription
        self._setup_marker_subscription()
        
        # Set up depth subscription
        self._setup_depth_subscription()
    
    # ---------- ROS COMMUNICATION SETUP ----------
    
    def _setup_marker_subscription(self):
        """Set up ROS subscription for object marker information"""

        
        self.object_marker_sub = rospy.Subscriber(
            self.object_mapper_topic,
            Marker,
            self.object_marker_callback
        )
    
    def _setup_depth_subscription(self):
        """Set up ROS subscription for depth information"""
        
        self.depth_sub = rospy.Subscriber(
            self.bbox_depth_topic,
            Float32MultiArray,
            self.depth_callback
        )
    
    # ---------- CALLBACK METHODS ----------
    
    def object_marker_callback(self, msg):
        """
        Process object marker messages
        
        Args:
            msg (Marker): Marker message from object detection
        """
        if self.scanning_in_progress:
            if msg.header.frame_id != "map":
                rospy.logwarn(f"Marker not in map frame: {msg.header.frame_id}")
                return
        
            # Store the validated marker
            self.object_marker = msg
    
    def depth_callback(self, msg):
        """
        Process depth information of detected objects
        
        Args:
            msg (Float32MultiArray): Message containing depth data
                Format: [n_boxes, cls_id, conf, x1, y1, x2, y2, depth]
        """
        data_length = len(msg.data)

        # Check for minimum data length required by new format
        if data_length >= 8:
            n_boxes = int(msg.data[0])
            if n_boxes < 1:
                rospy.logdebug("No boxes detected in depth message")
                return
             # Extract depth from the new format - position 7
            self.current_depth = float(msg.data[7])

            # Extract class ID from the new format - position 1
            self.current_class_id = int(msg.data[1])
                
    
    # ---------- STATE MANAGEMENT METHODS ----------
    
    def reset_detection_state(self):
        """Reset object detection state"""
        self.object_detected = False
        self.object_marker = None
    
    def get_remaining_angles(self):
        """
        Get the list of remaining scan angles
        
        Returns:
            list: Remaining angles to scan in degrees
        """
        return self.remaining_angles
    
    # ---------- DETECTION METHODS ----------
    
    def _scan_with_sustained_detection(self, angle, desired_class_id=None):
        """
        Scan for objects at current position with sustained detection requirement
        
        Requires that an object is continuously detected for a minimum duration
        to filter out false positives and ensure stable detection.
        
        Args:
            angle (float): Current rotation angle for logging
            desired_class_id (int, optional): Class ID to scan for, or None to accept any class
            
        Returns:
            bool: True if object detected with sustained requirement, False otherwise
        """
        rospy.loginfo(f"Starting object detection at {angle} degrees...")
        
        # Variables for sustained detection
        detection_start = None
        scan_start = rospy.Time.now()
        
        # Continue scanning until timeout
        while (rospy.Time.now() - scan_start) < self.scan_timeout and not rospy.is_shutdown():
                
            # Check if we have a marker
            if self.object_marker is not None:
                # Accept any class ID if desired_class_id is None, otherwise check for match
                class_match = (desired_class_id is None) or (self.current_class_id == desired_class_id)
                
                if class_match:
                    rospy.loginfo_throttle(5.0, f"Detected object with class ID {self.current_class_id}")

                    # Start or continue timing the sustained detection
                    if detection_start is None:
                        detection_start = rospy.Time.now()
                        rospy.loginfo("Potential object detected, timing sustained detection...")
                    else:
                        # Check if we've maintained detection long enough
                        elapsed = rospy.Time.now() - detection_start
                        if elapsed >= self.required_detection_duration:
                            # Validate detection with depth check
                            if self.current_depth is not None and self.current_depth < self.max_detection_depth:
                                self.object_detected = True
                                rospy.loginfo(f"Object detected at {angle} with class id {self.current_class_id}")
                                return True          
                else:
                    # Detected object does not match desired class, reset timer
                    rospy.logwarn_throttle(2.0, f"Detected object with incorrect class ID {self.current_class_id}")
                    detection_start = None
            else:
                # Lost detection, reset the timer
                if detection_start is not None:
                    rospy.logwarn("Lost detection during sustainment period, resetting timer")
                    detection_start = None      
            rospy.sleep(0.1)  # Check at 10Hz
        
        rospy.loginfo(f"No sustained detection at {angle} degrees within timeout")
        return False
    
    # ---------- ROTATION METHODS ----------
    
    def _perform_single_rotation_scan(self, current_pose, angle, rotation_index, total_rotations, desired_class_id):
        """
        Perform a single rotation and scan for objects
        
        1. Rotates to the specified angle
        2. Waits for camera stabilization
        3. Scans for objects with sustained detection requirement
        
        Args:
            current_pose: Current robot pose
            angle (float): Rotation angle in degrees
            rotation_index (int): Index of the current rotation
            total_rotations (int): Total number of rotations
            
        Returns:
            ScanResult: Result of the scan operation
        """
        # Reset object detection status for this rotation
        self.object_marker = None
        self.object_detected = False
        
        # Calculate orientation quaternion
        yaw = math.radians(angle)
        quaternion = quaternion_from_euler(0, 0, yaw)
        
        # Create goal at current position with new orientation
        current_goal = self.nav_controller.create_navigation_goal(
            current_pose.position.x,
            current_pose.position.y,
            Quaternion(*quaternion)
        )
        
        # Send rotation goal
        rospy.loginfo(f"Executing rotation {rotation_index+1}/{total_rotations} ({angle} degrees)")
        self.nav_controller.client.send_goal(current_goal)
        
        # Wait for rotation to complete with timeout
        if not self.nav_controller.client.wait_for_result(self.rotation_timeout):
            rospy.logwarn(f"Rotation timeout reached for angle {angle}")
            return ScanResult.ERROR
        
        if self.nav_controller.client.get_state() != actionlib.GoalStatus.SUCCEEDED:
            rospy.logwarn(f"Rotation goal failed for angle {angle}")
            return ScanResult.ERROR
        
        # Rotation complete, wait for camera to stabilize
        rospy.loginfo(f"Rotation to {angle} degrees complete, waiting {self.scan_stabilization_time}s for stabilization...")
        rospy.sleep(self.scan_stabilization_time)
        
        # Clear any previous detection that might be from motion blur
        self.object_marker = None
        
        # Scan for objects with sustained detection requirement
        detected = self._scan_with_sustained_detection(angle, desired_class_id)
        
        # Convert boolean result to ScanResult enum
        return ScanResult.OBJECT_DETECTED if detected else ScanResult.NO_DETECTION
    
    # ---------- MAIN SCANNING METHOD ----------
    
    def perform_scan_rotation(self, rotation_angles=None, desired_class_id=None, complete_scan=False):
        """
        Perform a sequence of rotations to scan for objects
        
        The robot will rotate to several orientations, pausing at each to allow
        for stabilized object detection with a sustained detection requirement.
        
        Args:
            rotation_angles (list, optional): List of angles in degrees to scan.
                                             If None, uses default angles.
            desired_class_id (int, optional): Class ID to scan for.
            complete_scan (bool, optional): If True, completes all rotations regardless
                                           of detections. If False, exits early on first detection.
        
        Returns:
            tuple: (ScanResult, remaining_angles) where:
                  - ScanResult is the result of the scanning operation
                  - remaining_angles is a list of angles not yet scanned
        """
        if rotation_angles is None:
            rotation_angles = self.rotation_angles.copy()
        
        rospy.loginfo(f"Starting scan rotation sequence with {len(rotation_angles)} angles: {rotation_angles}")
        rospy.loginfo(f"Scan mode: {'Complete scan' if complete_scan else 'Early exit on detection'}")
        
        # Set scanning flag to true and initialize state
        self.scanning_in_progress = True
        self.object_detected = False    
        self.remaining_angles = rotation_angles.copy()
        found_object = False
        
        # Get current position
        current_pose = self.nav_controller.get_robot_pose()
        if current_pose is None:
            rospy.logerr("Failed to get current robot pose")
            self.scanning_in_progress = False
            return ScanResult.ERROR, self.remaining_angles
        
        # Perform the rotations one by one
        for i in range(len(rotation_angles)):
            # Get the current angle to process (always the first in remaining list)
            if not self.remaining_angles:
                rospy.logwarn("No remaining angles to scan")
                break
                
            current_angle = self.remaining_angles[0]
            rospy.loginfo(f"Processing angle {i+1}/{len(rotation_angles)}: {current_angle} degrees")
            
            # Perform the scan at this angle
            result = self._perform_single_rotation_scan(
                current_pose, current_angle, i, len(rotation_angles), desired_class_id
            )
            
            # Remove this angle from remaining angles AFTER processing it
            self.remaining_angles.pop(0)
            rospy.loginfo(f"Remaining angles after processing {current_angle} degrees: {self.remaining_angles}")
            
            # If object detected, either return early or continue based on complete_scan flag
            if result == ScanResult.OBJECT_DETECTED:
                found_object = True
                rospy.loginfo(f"Object detected during scan rotation at angle {current_angle}")
                if not complete_scan:
                    self.scanning_in_progress = False
                    return ScanResult.OBJECT_DETECTED, self.remaining_angles
                # If complete_scan is True, continue to next angle
            
            # If error occurred, log and continue to next angle
            if result == ScanResult.ERROR:
                rospy.logwarn(f"Error during scan at angle {current_angle}, continuing to next angle")
            
            # Brief pause between rotations
            if i < len(rotation_angles) - 1:  # Don't pause after the last rotation
                rospy.sleep(0.5)
        
        # End of rotation sequence
        self.scanning_in_progress = False
        
        if found_object:
            rospy.loginfo("Completed scan rotation sequence, detected objects at one or more angles")
            return ScanResult.OBJECT_DETECTED, self.remaining_angles
        else:
            rospy.loginfo("Completed scan rotation sequence, no objects detected")
            return ScanResult.NO_DETECTION, self.remaining_angles


# ---------- DIRECT EXECUTION BLOCK ----------

if __name__ == "__main__":
    try:
        # This class requires a NavigationController, so we can't run it directly
        rospy.loginfo("ObjectScanner cannot be run directly as it requires a NavigationController")
    except rospy.ROSInterruptException:
        rospy.loginfo("Object scanner interrupted")

File: src/object_scout_cpp/src/object_scanner.cpp
#include "object_scout_cpp/object_scanner.hpp"
#include <cmath>

ObjectScanner::ObjectScanner(
    const std::string& robot_name,
    NavigationController* nav_controller,
    const std::string& object_mapper_topic,
    const std::string& bbox_depth_topic
) :
    robot_name_(robot_name),
    nav_controller_(nav_controller),
    object_mapper_topic_(object_mapper_topic),
    bbox_depth_topic_(bbox_depth_topic),
    scan_stabilization_time_(1.0),
    required_detection_duration_(2.0),
    scan_timeout_(3.0),
    rotation_timeout_(45.0),
    max_detection_depth_(10.0),
    scanning_in_progress_(false),
    object_detected_(false),
    current_depth_(0.0),
    current_class_id_(-1)
{
    // Initialize default rotation angles
    default_rotation_angles_ = {0, 45, 90, 135, 180, 225, 270, 315};
    
    // Setup ROS subscribers
    ros::NodeHandle nh;
    
    // Object marker subscription
    object_marker_sub_ = nh.subscribe(
        object_mapper_topic_,
        1,
        &ObjectScanner::objectMarkerCallback,
        this
    );
    
    // Depth subscription
    depth_sub_ = nh.subscribe(
        bbox_depth_topic_,
        1,
        &ObjectScanner::depthCallback,
        this
    );
    
    ROS_INFO_STREAM("ObjectScanner initialized for robot: " << robot_name_);
}

void ObjectScanner::resetDetectionState() {
    object_detected_ = false;
    object_marker_.reset();
}

std::vector<double> ObjectScanner::getRemainingAngles() const {
    return remaining_angles_;
}

void ObjectScanner::objectMarkerCallback(const visualization_msgs::Marker::ConstPtr& msg) {
    if (scanning_in_progress_) {
        if (msg->header.frame_id != "map") {
            ROS_WARN_STREAM("Marker not in map frame: " << msg->header.frame_id);
            return;
        }
        
        // Store the validated marker
        object_marker_ = msg;
    }
}

void ObjectScanner::depthCallback(const std_msgs::Float32MultiArray::ConstPtr& msg) {
    size_t data_length = msg->data.size();
    
    // Check for minimum data length required by new format
    if (data_length >= 8) {
        int n_boxes = static_cast<int>(msg->data[0]);
        if (n_boxes < 1) {
            ROS_DEBUG("No boxes detected in depth message");
            return;
        }
        
        // Extract depth from the new format - position 7
        current_depth_ = msg->data[7];
        
        // Extract class ID from the new format - position 1
        current_class_id_ = static_cast<int>(msg->data[1]);
    }
}

bool ObjectScanner::scanWithSustainedDetection(double angle, int desired_class_id) {
    ROS_INFO_STREAM("Starting object detection at " << angle << " degrees...");
    
    // Variables for sustained detection
    ros::Time detection_start;
    bool has_detection_start = false;
    ros::Time scan_start = ros::Time::now();
    
    // Continue scanning until timeout
    while ((ros::Time::now() - scan_start) < scan_timeout_ && ros::ok()) {
        // Check if we have a marker
        if (object_marker_) {
            // Accept any class ID if desired_class_id is -1, otherwise check for match
            bool class_match = (desired_class_id == -1) || (current_class_id_ == desired_class_id);
            
            if (class_match) {
                ROS_INFO_STREAM_THROTTLE(5.0, "Detected object with class ID " << current_class_id_);
                
                // Start or continue timing the sustained detection
                if (!has_detection_start) {
                    detection_start = ros::Time::now();
                    has_detection_start = true;
                    ROS_INFO("Potential object detected, timing sustained detection...");
                } else {
                    // Check if we've maintained detection long enough
                    ros::Duration elapsed = ros::Time::now() - detection_start;
                    if (elapsed >= required_detection_duration_) {
                        // Validate detection with depth check
                        if (current_depth_ > 0 && current_depth_ < max_detection_depth_) {
                            object_detected_ = true;
                            ROS_INFO_STREAM("Object detected at " << angle << " with class id " << current_class_id_);
                            return true;
                        }
                    }
                }
            } else {
                // Detected object does not match desired class, reset timer
                ROS_WARN_STREAM_THROTTLE(2.0, "Detected object with incorrect class ID " << current_class_id_);
                has_detection_start = false;
            }
        } else {
            // Lost detection, reset the timer
            if (has_detection_start) {
                ROS_WARN("Lost detection during sustainment period, resetting timer");
                has_detection_start = false;
            }
        }
        
        ros::Duration(0.1).sleep();  // Check at 10Hz
    }
    
    ROS_INFO_STREAM("No sustained detection at " << angle << " degrees within timeout");
    return false;
}

ScanResult ObjectScanner::performSingleRotationScan(
    const geometry_msgs::Pose& current_pose,
    double angle,
    int rotation_index,
    int total_rotations,
    int desired_class_id
) {
    // Reset object detection status for this rotation
    object_marker_.reset();
    object_detected_ = false;
    
    // Calculate orientation quaternion
    double yaw = angle * M_PI / 180.0;  // Convert to radians
    tf2::Quaternion tf_quat;
    tf_quat.setRPY(0, 0, yaw);
    geometry_msgs::Quaternion orientation = tf2::toMsg(tf_quat);
    
    // Send rotation goal
    ROS_INFO_STREAM("Executing rotation " << (rotation_index + 1) << "/" << total_rotations << " (" << angle << " degrees)");
    
    // Attempt to rotate using navigation controller
    bool rotation_success = nav_controller_->moveToPosition(
        current_pose.position.x,
        current_pose.position.y,
        &orientation,
        rotation_timeout_.toSec()
    );
    
    if (!rotation_success) {
        ROS_WARN_STREAM("Rotation goal failed for angle " << angle);
        return ScanResult::ERROR;
    }
    
    // Rotation complete, wait for camera to stabilize
    ROS_INFO_STREAM("Rotation to " << angle << " degrees complete, waiting " << scan_stabilization_time_ << "s for stabilization...");
    ros::Duration(scan_stabilization_time_).sleep();
    
    // Clear any previous detection that might be from motion blur
    object_marker_.reset();
    
    // Scan for objects with sustained detection requirement
    bool detected = scanWithSustainedDetection(angle, desired_class_id);
    
    // Convert boolean result to ScanResult enum
    return detected ? ScanResult::OBJECT_DETECTED : ScanResult::NO_DETECTION;
}

std::pair<ScanResult, std::vector<double>> ObjectScanner::performScanRotation(
    const std::vector<double>& rotation_angles,
    int desired_class_id,
    bool complete_scan
) {
    // Use default angles if none provided
    std::vector<double> angles_to_use;
    if (rotation_angles.empty()) {
        angles_to_use = default_rotation_angles_;
    } else {
        angles_to_use = rotation_angles;
    }
    
    ROS_INFO_STREAM("Starting scan rotation sequence with " << angles_to_use.size() << " angles");
    ROS_INFO_STREAM("Scan mode: " << (complete_scan ? "Complete scan" : "Early exit on detection"));
    
    // Set scanning flag to true and initialize state
    scanning_in_progress_ = true;
    object_detected_ = false;
    remaining_angles_ = angles_to_use;
    bool found_object = false;
    
    // Get current position
    geometry_msgs::Pose current_pose = nav_controller_->getRobotPose();
    
    // Perform the rotations one by one
    for (size_t i = 0; i < angles_to_use.size(); ++i) {
        // Get the current angle to process (always the first in remaining list)
        if (remaining_angles_.empty()) {
            ROS_WARN("No remaining angles to scan");
            break;
        }
        
        double current_angle = remaining_angles_[0];
        ROS_INFO_STREAM("Processing angle " << (i + 1) << "/" << angles_to_use.size() << ": " << current_angle << " degrees");
        
        // Perform the scan at this angle
        ScanResult result = performSingleRotationScan(
            current_pose, 
            current_angle, 
            i, 
            angles_to_use.size(),
            desired_class_id
        );
        
        // Remove this angle from remaining angles AFTER processing it
        remaining_angles_.erase(remaining_angles_.begin());
        ROS_INFO_STREAM("Remaining angles after processing " << current_angle << " degrees: " << remaining_angles_.size());
        
        // If object detected, either return early or continue based on complete_scan flag
        if (result == ScanResult::OBJECT_DETECTED) {
            found_object = true;
            ROS_INFO_STREAM("Object detected during scan rotation at angle " << current_angle);
            if (!complete_scan) {
                scanning_in_progress_ = false;
                return std::make_pair(ScanResult::OBJECT_DETECTED, remaining_angles_);
            }
            // If complete_scan is true, continue to next angle
        }
        
        // If error occurred, log and continue to next angle
        if (result == ScanResult::ERROR) {
            ROS_WARN_STREAM("Error during scan at angle " << current_angle << ", continuing to next angle");
        }
        
        // Brief pause between rotations
        if (i < angles_to_use.size() - 1) {  // Don't pause after the last rotation
            ros::Duration(0.5).sleep();
        }
    }
    
    // End of rotation sequence
    scanning_in_progress_ = false;
    
    if (found_object) {
        ROS_INFO("Completed scan rotation sequence, detected objects at one or more angles");
        return std::make_pair(ScanResult::OBJECT_DETECTED, remaining_angles_);
    } else {
        ROS_INFO("Completed scan rotation sequence, no objects detected");
        return std::make_pair(ScanResult::NO_DETECTION, remaining_angles_);
    }
}